### ISSUE TITLE
I did not found yolov8s.yaml file where do i get this file ? Then how to do i do the model tuning and how to change different function parameter such as activation, filter and layer, and output? 

### ISSUE THREAD

**Issue by jahid-coder** [ultralytics/ultralytics#5757]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

Recently, I have used yolov8s.pt to detect object . But I want to model tuning and want to change different function parameter
such as activation, filter and layer, and output.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by jahid-coder** [comment]
@glenn-jocher 
Thanks for your response, But yet i can't find yolov8s.yaml file . where do i get this. i share the below image  . Here, i can't see such file. Please tell me exact location where i find yolov8s.yaml file. I am confuse about this.
![image](https://github.com/ultralytics/ultralytics/assets/73584067/0a043a85-7a79-4dfa-b74e-6110af694140)

**Comment by glenn-jocher** [comment]
yolov8.yaml is what you want, this generates all 5 sizes when called as ie yolov8s.yaml

**Comment by jahid-coder** [comment]
@glenn-jocher  
thanks for your response, please see my below image and code , here how can i set up model activation , filter size and output parameter to modify model those parameter.
![image](https://github.com/ultralytics/ultralytics/assets/73584067/f92008cf-152e-4a42-bb9f-97eed477d764)

**Comment by glenn-jocher** [comment]
@jahid-coder hello,

Thanks for your message. In the YAML model architecture file, there are various fields to customize the layer in the model. This file describes the architecture of the model, including parameters like activation functions, filters, and layers.

**Setting up the model activation:** You can find the field named 'activation' in the YAML file, where you can replace the current activation function with the one you need. Available functions include 'leaky', 'swish', 'mish', and others. Please make sure to use an activation function that is compatible with your architecture and task.

**Filter size:** Similarly, the size of the filters (the number of output channels, in CNN parlance) can also be configured in this YAML file. This is usually denoted by the 'filters' word in the file. 

**Output parameter:** If you're referring to the final layer output, then it is represented by the number of output channels in the final Convolution block, this can be determined by the number of object classes you're training the model on, and is directly dependent on the dataset you use for training.

Remember, model tuning requires a good understanding of the architecture of the YOLO models as well as Convolutional Neural Networks in general. Any changes should be made with care to preserve the input-output relationships between layers in the architecture.

Hope this helps, and all the best in your adaptations.

Best,
Glenn

**Comment by jahid-coder** [comment]
@glenn-jocher hello,
thanks for response. But yet i can't understand how to modify this below code  to set up model activation , filter size and output parameter in here .I can't understand this code . Is there any documentation to understand this code. In colab., how to modify yolov8.yaml file using CLI command. 

# Ultralytics YOLO 🚀, AGPL-3.0 license
# YOLOv8 object detection model with P3-P5 outputs. For Usage examples see https://docs.ultralytics.com/tasks/detect

# Parameters
nc: 80  # number of classes
scales: # model compound scaling constants, i.e. 'model=yolov8n.yaml' will call yolov8.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.33, 0.25, 1024]  # YOLOv8n summary: 225 layers,  3157200 parameters,  3157184 gradients,   8.9 GFLOPs
  s: [0.33, 0.50, 1024]  # YOLOv8s summary: 225 layers, 11166560 parameters, 11166544 gradients,  28.8 GFLOPs
  m: [0.67, 0.75, 768]   # YOLOv8m summary: 295 layers, 25902640 parameters, 25902624 gradients,  79.3 GFLOPs
  l: [1.00, 1.00, 512]   # YOLOv8l summary: 365 layers, 43691520 parameters, 43691504 gradients, 165.7 GFLOPs
  x: [1.00, 1.25, 512]   # YOLOv8x summary: 365 layers, 68229648 parameters, 68229632 gradients, 258.5 GFLOPs

# YOLOv8.0n backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]]  # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]]  # 1-P2/4
  - [-1, 3, C2f, [128, True]]
  - [-1, 1, Conv, [256, 3, 2]]  # 3-P3/8
  - [-1, 6, C2f, [256, True]]
  - [-1, 1, Conv, [512, 3, 2]]  # 5-P4/16
  - [-1, 6, C2f, [512, True]]
  - [-1, 1, Conv, [1024, 3, 2]]  # 7-P5/32
  - [-1, 3, C2f, [1024, True]]
  - [-1, 1, SPPF, [1024, 5]]  # 9

# YOLOv8.0n head
head:
  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 6], 1, Concat, [1]]  # cat backbone P4
  - [-1, 3, C2f, [512]]  # 12

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 4], 1, Concat, [1]]  # cat backbone P3
  - [-1, 3, C2f, [256]]  # 15 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P4
  - [-1, 3, C2f, [512]]  # 18 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P5
  - [-1, 3, C2f, [1024]]  # 21 (P5/32-large)

  - [[15, 18, 21], 1, Detect, [nc]]  # Detect(P3, P4, P5)

**Comment by glenn-jocher** [comment]
@jahid-coder hello,

The YAML file lists the architectural components of the model and their corresponding parameters. It enables you to control how the model is constructed. Here's how you can interpret the structure:

- The `scales` section allows you to choose different pre-set scales for your model (e.g., `n`, `s`, `m`, `l`, `x`). Each scale has different `depth`, `width`, `max_channels` which in turn influence the model size, and thus accuracy and speed trade-off.

- The `backbone` section is used to define the architecture of the backbone network. It's essentially a sequence of blocks with each block like `- [-1, 1, Conv, [64, 3, 2]]` representing a series of operations. The first integer is the layer from which the input comes, the second integer stands for the number of times this block is repeated. The string represents the type of layer (e.g., Convolution, MaxPooling, etc.) and the list in square brackets are the parameters for those layers.

- The `head` section is structured similarly and used to define the architecture of the head network, which processes the features extracted by the backbone and makes final predictions.

Regarding your specific questions:

- To change the model activation, you need to look at where the activation function is specified (usually in the form of a layer type like 'Mish', 'LeakyReLU'), and replace it with your desired activation function.

- To change the filter sizes, you need to modify the list of parameters for the convolutional layers `Conv`. The first value in the list (e.g., [64, 3, 2]) is the number of filters.

- To modify the `output` parameters of the model, alter the final layers in the `head` section. Specifically, the number of output channels in the final Convolution block should equal the number of classes you're training the model on. This is represented by `nc` in the `Detect` layer.

As for Colab commands to modify the file, you can't directly use Command Line Interface commands in a Colab notebook to modify files. You can, however, read in the file using Python, modify its content, and then write it back to the same or a new file.

I hope this explanation helps you grasp how to alter the model architecture using the YAML file. Remember to proceed with caution and ensure any amendments

**Comment by jahid-coder** [comment]
@glenn-jocher  hello,
thanks for your response. Now i feel batter understand about the the above yaml file . But yet i don't know  how to modify this file using python code . Have any documentation that help me and guide to me , to modify this file and getting good output.

**Comment by glenn-jocher** [comment]
Hello @jahid-coder,

Glad to hear that you have a better understanding now. In regards to modifying the YAML file using Python code, yes it is possible. Here are some steps you can follow without providing the exact code:

1. First, you need to read the YAML file into Python. Python's built-in module known as 'PyYAML' allows you to parse YAML files into Python dictionaries giving you the ability to access and modify the values.

2. To update parameters within this file, you can then manipulate the resulting Python dictionary. For example, you can navigate to the 'head' section which is a list of lists in the dictionary and alter the values as needed.

3. When you're done, you will need to write your changes back to the YAML file. PyYAML provides a way to dump Python dictionaries back into YAML format, which you can then write into the file.

Regarding the error shown in your screenshots, it appears that you're trying to access values in a dictionary using a key that doesn't exist in the dictionary. This normally happens when you're trying to access a key-value pair in a nested dictionary but the parent dictionary doesn't contain the parent key. 

Please make sure the keys you're using are available in the dictionary. To be safer, you can also use the `dict.get(key)` method. If the key doesn’t exist this method returns `None`, therefore it won't throw an error.

Please check if this resolves your issue. If not, don't hesitate to ask for further clarification.

Best of luck!
Glenn

**Comment by jahid-coder** [comment]
Hello @glenn-jocher, 
thanks for your lucid response, I get it and i can modified the file and save the file . Now , i have  question that after modified the yolov8.yaml file , how can use it like as yolov8s.pt model . please explain step by step that will be helpful to me. thanks

**Comment by glenn-jocher** [comment]
Hello @jahid-coder,

Great to hear that you were able to modify and save the YAML file! 

To your questions:

1. Activation Function: The activation function is typically specified within the definition of each layer in your model architecture (Conv, C2f, etc.). However, in this specific YAML configuration, the activation function is not explicitly stated and is controlled in the code where each of these module is defined.

2. Convolutional Layers: You can change the filter sizes of any Conv layer or all of them based on your requirements. However, changes should be made with consideration to preserve the input-output relationships between layers in the architecture.

After modifying the yolov8.yaml file, to use it in a similar fashion to yolov8s.pt model, you would:

1. Load the modified YAML to construct the new model architecture.
2. Train this new model architecture on your dataset via any defined training process, until the performance is satisfactory.
3. Save the trained model weights.
4. Now you can load and use these saved weights just like you would use a .pt file, for inference or further training.

Remember, the .pt file not only contains the architecture defined in the YAML, but also the trained model weights. So, after modifying the YAML file, you will need to retrain the model to generate a new .pt file with updated architecture.

Hope this helps! Let me know if you have more questions.

**Comment by jahid-coder** [comment]
Hello @glenn-jocher ,
thanks for your effective message.  I am facing a problem that to Convert the modified dictionary back to YAML format using yaml.dump(loaded_data, default_flow_style=False)  then i got only the backbone and head section why do i not get the code before backbone section . see the below image 

![image](https://github.com/ultralytics/ultralytics/assets/73584067/be3ad2ee-02a3-4b28-aa75-7c4532f87aa3)

In the above image, when i try to convert the modified dictionary back to YAML format using  yaml.dumloaded_data, default_flow_style=False), the underline of the section of below image is omitted . I got only the backbone and head section. please tell me what's the problem here .
![image](https://github.com/ultralytics/ultralytics/assets/73584067/5755e16d-4195-466c-8ba8-355a0cd35fb3)

**Comment by glenn-jocher** [comment]
Hello @jahid-coder,

From your description, it sounds like there may be an inconsistency between the original structure of the YAML file and how it's being translated into a dictionary then back to YAML.

When parsing the YAML file, ensure that all sections - not just 'backbone' and 'head', but also those early sections - are correctly decoded and stored in the dictionary. These sections should be represented as key-value pairs in the Python dictionary.

While it's hard to diagnose without seeing the exact sequence of operations you're performing, be aware that any modifications you make that accidentally drop or overlook keys could lead to parts of the file not being included when you convert back to YAML.

Remember to be meticulous when accessing and modifying nested structures within the dictionary. If you've expectedly removed or renamed keys within the dictionary, this could explain why certain parts are absent when dumping back to YAML format.

Hopefully this gives you some pointers on what to look out for.
